node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -107,46 +76,235 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
 		})
 		.collect::<Vec<_>>();
 
-	let generics = quote! {
-		< S, #( #baggage_generic_ty, )* >
-	};
-	let where_clause = quote! {
+	// Helpers to use within quote! macros
+	let builder_where_clause = quote! {
 		where
-			S: #support_crate ::SpawnNamed,
+			#(
+				#subsystem_type : Subsystem<#subsystem_ctx_name< #consumes >, #error_ty>,
+			)*
+			S: #support_crate ::SpawnNamed + Send
 	};
-
 	let builder_generics = quote! {
-		<S, #( #baggage_generic_ty, )* #( #builder_generic_ty, )* >
+		S, #( #subsystem_type, )* #( #baggage_generic_ty, )*
 	};
 
-	// all subsystems must have the same context
-	// even if the overseer does not impose such a limit.
-	let builder_additional_generics = quote! {
-		<#( #builder_generic_ty, )* >
-	};
+	let field_name = subsystem_name.iter().chain(baggage_name.iter()).collect::<Vec<_>>();
+	let field_type = subsystem_type
+		.iter()
+		.chain(info.baggage().iter().filter_map(|bag| bag.field_ty.get_ident()))
```

Comment:
This should not be a `filter_map`, `x.field_ty.get_ident()` being `None` is a bug, that should not cause broken code to be generated. We only support named fields for the overseer pattern. If it's handled earlier, it should be a `expect`.

Reply:
I have changed it in https://github.com/paritytech/polkadot/pull/4753/commits/0af20ce752fc228314f5ec37a44e9943dcd78ab0

With `filter_map` approach we miss all complex fields (like `Vec<T>`)